aten/src/ATen/native/cpu/BinaryOpsKernel.cpp

```diff
@@ -56,6 +56,12 @@ void atan2_kernel(TensorIteratorBase& iter) {
 void mul_kernel(TensorIteratorBase& iter) {
   if (iter.dtype() == ScalarType::Bool) {
     cpu_kernel(iter, [=](bool a, bool b) -> bool { return a && b; });
+  } else if (iter.dtype() == kComplexHalf) {
+    using scalar_t = c10::complex<at::Half>;
+    cpu_kernel(iter, [=](scalar_t a, scalar_t b) -> scalar_t {
```

Comment:
Supporting `cpu_kernel_vec` required supporting `operator +=` on `complex<Half>`. I think it might be ok to have non-vectorised kernel for now.

Reply:
should we file an issue and a comment for future reference?

Reply:
Ugh, why is cpu doing intermediate math in low precision type? They are doing it everywhere, and it's wrong

Reply:
Also what's the issue with supporting `+=` for `complex<Half>`?

Reply:
> Ugh, why is cpu doing intermediate math in low precision type? They are doing it everywhere, and it's wrong

We are doing it higher precision right ? `complex<float>` and casting the result back.

> Also what's the issue with supporting += for complex<Half>

I think ideally `complex<Half>` should be a minimal data-storage structure with almost no math operator support and the kernel supporting it should take care of upcasting, compute in upcasted type and then explicitly downcasting.

Reply:
For `cpu` we are doing that right now for sure.

https://github.com/pytorch/pytorch/blob/d05eaa2a4200d2a9f595836ea0e749ffb6e24eb7/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L60-L64

Reply:
Sorry I meant implement `*=` not `+=`

Reply:
uhh is it being used at the end ...? `static_cast<scalar_t>(...)`

Reply:
Yes. `scalar_t` is `complex<Half>`. So for each element a and b, we upcast them to `comp_t` which is `complex<float>`, compute the multiplication and then downcast explicitly back to `complex<Half>`. ( Should get rid of scalar_t and comp_t and use raw types to make it more clear, wdyt?)

Reply:
But that means that cpu_kernel_vec needs support for converting a vec of complex<Half> into a vec of complex<Float> and back, and not *= on complex<Half>?

Reply:
So the calls go as follow. We load the generic `Vec<ComplexHalf>` from `vec_base.h` (Note that this generic struct does not use vector instruction but loops over and maps the operator.
https://github.com/pytorch/pytorch/blob/1a7e43be141ce01469d7605075cb1008bf19abd7/aten/src/ATen/cpu/vec/vec_base.h#L95

`operator*` of this Sec  is called
https://github.com/pytorch/pytorch/blob/1a7e43be141ce01469d7605075cb1008bf19abd7/aten/src/ATen/cpu/vec/vec_base.h#L570-L576

which calls into `operator*` for `complex<Half>` which matches following template function
https://github.com/pytorch/pytorch/blob/1a7e43be141ce01469d7605075cb1008bf19abd7/c10/util/complex.h#L379-L383

Note that this expects `*=` to be implemented.